source/pki_utils.c

```diff
@@ -1,298 +1,816 @@
+        aws_byte_buf_clean_up(&aggregate_buffer);
+        return AWS_OP_ERR;
+    }
+
+    CFArrayRef import_output;
+    SecExternalFormat format = kSecFormatUnknown;
+    SecExternalItemType item_type = kSecItemTypeAggregate;
+
+    SecItemImportExportKeyParameters import_params;
+    AWS_ZERO_STRUCT(import_params);
+    import_params.version = SEC_KEY_IMPORT_EXPORT_PARAMS_VERSION;
+    import_params.passphrase = CFSTR("");
+
+    SecKeychainRef import_keychain = NULL;
+    SecKeychainCopyDefault(&import_keychain);
+
+    OSStatus status = SecItemImport(
+        aggregate_certificate_data, NULL, &format, &item_type, 0, &import_params, import_keychain, &import_output);
+
+    CFRelease(aggregate_certificate_data);
+    aws_byte_buf_clean_up(&aggregate_buffer);
+
+    if (status != errSecSuccess && status != errSecDuplicateItem) {
+        return AWS_OP_ERR;
+    }
+
+    SecCertificateRef certificate_ref = NULL;
+
+    /* if it's already there, just conver this over to a cert and then let the keychain give it back to us. */
+    if (status == errSecDuplicateItem) {
+        struct aws_array_list cert_chain_list;
+
+        if (aws_array_list_init_dynamic(&cert_chain_list, alloc, 2, sizeof(struct aws_byte_buf))) {
+            return AWS_OP_ERR;
+        }
+
+        if (aws_decode_pem_to_buffer_list(alloc, public_cert_chain, &cert_chain_list)) {
+            aws_array_list_clean_up(&cert_chain_list);
+            return AWS_OP_ERR;
+        }
+
+        struct aws_byte_buf *root_cert_ptr = NULL;
+        aws_array_list_get_at_ptr(&cert_chain_list, (void **)&root_cert_ptr, 0);
+        assert(root_cert_ptr);
+        CFDataRef root_cert_data = CFDataCreate(cf_alloc, root_cert_ptr->buffer, root_cert_ptr->len);
+
+        if (root_cert_data) {
+            certificate_ref = SecCertificateCreateWithData(cf_alloc, root_cert_data);
+            CFRelease(root_cert_data);
+        }
+
+        aws_cert_chain_clean_up(&cert_chain_list);
+        aws_array_list_clean_up(&cert_chain_list);
+    } else {
+        certificate_ref = (SecCertificateRef)CFArrayGetValueAtIndex(import_output, 0);
+        /* SecCertificateCreateWithData returns an object with +1 retain, so we need to match that behavior here */
+        CFRetain(certificate_ref);
+    }
+
+    if (certificate_ref) {
+        SecIdentityRef identity_output;
+        bool cleanup_import_output = status != errSecDuplicateItem;
+        status = SecIdentityCreateWithCertificate(import_keychain, certificate_ref, &identity_output);
+
+        CFRelease(certificate_ref);
+        CFRelease(import_keychain);
+        if (import_output && cleanup_import_output) {
+            CFRelease(import_output);
+        }
+
+        if (status == errSecSuccess) {
+            CFTypeRef certs[] = {identity_output};
+            *identity = CFArrayCreate(cf_alloc, (const void **)certs, 1L, &kCFTypeArrayCallBacks);
+            return AWS_OP_SUCCESS;
+        }
+    } else {
+        CFRelease(import_keychain);

```

Comment:
Doesn't the import_output still need to be cleaned up here?

Reply:
no, it's ref counted and gets picked up in the clean up code.